### PR TITLE
Use sync instead of stdbuf

### PR DESCRIPTION
**Pull request by haorenfsa**
Signed-off-by: shaoyue.chen <shaoyue.chen@zilliz.com>

**Description of the change**

- use `sync` instead of `stdbuf`

**Benefits**

- Reduce the chances that `member_id` & `member_removal.log` files are not being persisted

**Possible drawbacks**

**Applicable issues**

#39

**Additional information**

### CODE REVIEW

File: 3.3/debian-10/rootfs/opt/bitnami/scripts/libetcd.sh [hunk 1]
```diff
@@ -306,19 +306,21 @@ etcdctl_auth_flags() {
...
     is_boolean_yes "$ETCD_ON_K8S" && extra_flags+=("--endpoints=$(etcdctl_get_endpoints)")
     if retry_while "etcdctl ${extra_flags[*]} member list" >/dev/null 2>&1; then
-        while [[ ! -s "${ETCD_DATA_DIR}/member_id" ]]; do
-            # We use 'stdbuf' to ensure memory buffers are flushed to disk
-            # so we reduce the chances that the "member_id" file is empty.
-            # ref: https://www.gnu.org/software/coreutils/manual/html_node/stdbuf-invocation.html#stdbuf-invocation
+        while is_empty_value "${member_id}"; do
             read -r -a advertised_array <<<"$(tr ',;' ' ' <<<"$ETCD_ADVERTISE_CLIENT_URLS")"
-            stdbuf -oL etcdctl "${extra_flags[@]}" member list | grep -w "${advertised_array[0]}" | awk -F "," '{ print $1}' >"${ETCD_DATA_DIR}/member_id" || true
+            member_id=etcdctl "${extra_flags[@]}" member list | grep -w "${advertised_array[0]}" | awk -F "," '{ print $1}' || true
```
Comment (joancafom) [inline 3.3/debian-10/rootfs/opt/bitnami/scripts/libetcd.sh:317]: We usually make use of sub-shells to assign values to existing variables. IMHO, it is easier to read.

```suggestion
            member_id="$(etcdctl "${extra_flags[@]}" member list | grep -w "${advertised_array[0]}" | awk -F "," '{ print $1}' || true)"
```

**Review by joancafom — CHANGES_REQUESTED** [review]
Thanks so much for your PR @haorenfsa ! I left you a couple of comments, could you please review them?

**Comment by haorenfsa** [comment]
> Thanks so much for your PR @haorenfsa ! I left you a couple of comments, could you please review them?

Thank you. It's updated.

**Comment by haorenfsa** [comment]
@joancafom All updated

**Comment by joancafom** [comment]
Hi again @haorenfsa 

Changes LGTM now. As this logic might seem trivial, I want to ensure that the use of `sync` does not come with side effects. I will test the changes internally and come back soon, either to provide further feedback or to give the green light and merge this. 

> Btw, what do you think about the other PR?

As regards this, I did not have time to review it yet. The changes there do not seem to be trivial, so in the meantime, I consider we may resort to this PR `sync` and see how that handles the file persistence issue.

File: 3.3/debian-10/rootfs/opt/bitnami/scripts/libetcd.sh [hunk 2]
```diff
@@ -306,19 +306,21 @@ etcdctl_auth_flags() {
...
     info "Obtaining cluster member ID"
     etcd_start_bg
     read -r -a extra_flags <<<"$(etcdctl_auth_flags)"
     is_boolean_yes "$ETCD_ON_K8S" && extra_flags+=("--endpoints=$(etcdctl_get_endpoints)")
     if retry_while "etcdctl ${extra_flags[*]} member list" >/dev/null 2>&1; then
-        while [[ ! -s "${ETCD_DATA_DIR}/member_id" ]]; do
-            # We use 'stdbuf' to ensure memory buffers are flushed to disk
-            # so we reduce the chances that the "member_id" file is empty.
-            # ref: https://www.gnu.org/software/coreutils/manual/html_node/stdbuf-invocation.html#stdbuf-invocation
+        while is_empty_value "${member_id}"; do
```
Comment (joancafom) [inline 3.3/debian-10/rootfs/opt/bitnami/scripts/libetcd.sh:315]:
```suggestion
        while is_empty_value "$member_id"; do
```

**Review by joancafom — CHANGES_REQUESTED** [review]
Some feedback coming from the testing phase 😁

File: 3.3/debian-10/rootfs/opt/bitnami/scripts/libetcd.sh [hunk 3]
```diff
@@ -306,19 +306,21 @@ etcdctl_auth_flags() {
 ########################
 etcd_store_member_id() {
     local -a extra_flags
-
+    local member_id
```
Comment (joancafom) [inline 3.3/debian-10/rootfs/opt/bitnami/scripts/libetcd.sh:309]: You need to assign it an initial value, else the script will fail with `member_id: unbound variable` on line 315.

```suggestion
    local member_id=""
```


You need to assign an initial value, else bash will report `member_id: unbound variable` on line 315.

**Comment by haorenfsa** [comment]
Updated

**Pull request merged**